File: python-fastapi-CRUD/benchmark.py
from __future__ import annotations

import os
import statistics
import sys
import time
import urllib.error
import urllib.request
from concurrent.futures import ThreadPoolExecutor, as_completed


def env_int(key: str, default: int) -> int:
    value = os.environ.get(key)
    if value and value.isdigit() and int(value) > 0:
        return int(value)
    return default


def fetch(url: str) -> tuple[float, bool]:
    start = time.perf_counter()
    try:
        with urllib.request.urlopen(url, timeout=30) as resp:
            resp.read()
            ok = 200 <= resp.status < 300
    except (urllib.error.URLError, TimeoutError):
        ok = False
    elapsed = time.perf_counter() - start
    return elapsed, ok


def percentile(sorted_values: list[float], p: int) -> float:
    if not sorted_values:
        return 0.0
    idx = min((p * len(sorted_values)) // 100, len(sorted_values) - 1)
    return sorted_values[idx]


def main() -> None:
    url = os.environ.get("BENCH_URL", "http://localhost:8000/students")
    total = env_int("BENCH_REQUESTS", 10000)
    concurrency = env_int("BENCH_CONCURRENCY", 50)

    print("Python benchmark client")
    print(f"  Target:      {url}")
    print(f"  Requests:    {total}")
    print(f"  Concurrency: {concurrency}\n")

    try:
        urllib.request.urlopen(url, timeout=5).read()
    except Exception as exc:
        print(f"Server not reachable at {url}: {exc}", file=sys.stderr)
        sys.exit(1)

    latencies: list[float] = []
    ok_count = 0
    failed = 0

    start = time.perf_counter()
    with ThreadPoolExecutor(max_workers=concurrency) as pool:
        futures = [pool.submit(fetch, url) for _ in range(total)]
        for future in as_completed(futures):
            elapsed, ok = future.result()
            latencies.append(elapsed)
            if ok:
                ok_count += 1
            else:
                failed += 1
    duration = time.perf_counter() - start

    latencies.sort()
    rps = total / duration if duration > 0 else 0

    print("Results")
    print("-------")
    print("Language: Python (FastAPI client)")
    print(f"URL: {url}")
    print(f"Duration: {duration:.3f}s")
    print(f"Requests/sec: {rps:.2f}")
    print(f"Successful: {ok_count}")
    print(f"Failed: {failed}")
    print(f"Avg latency: {statistics.mean(latencies) * 1000:.2f}ms")
    print(f"p50 latency: {percentile(latencies, 50) * 1000:.2f}ms")
    print(f"p95 latency: {percentile(latencies, 95) * 1000:.2f}ms")
    print(f"p99 latency: {percentile(latencies, 99) * 1000:.2f}ms")


if __name__ == "__main__":
    main()

File: typescript-express-CRUD/src/benchmark.ts
/**
 * Load-test client for the Express CRUD server.
 *
 * Run the server first:
 *   npm run build && npm start
 *
 * Then:
 *   npm run benchmark
 *
 * Environment variables:
 *   BENCH_URL          (default: http://localhost:3000/students)
 *   BENCH_REQUESTS     (default: 10000)
 *   BENCH_CONCURRENCY  (default: 50)
 */

const DEFAULT_URL = "http://localhost:3000/students";
const DEFAULT_REQUESTS = 10_000;
const DEFAULT_CONCURRENCY = 50;

function envInt(key: string, fallback: number): number {
  const value = process.env[key];
  if (value && /^\d+$/.test(value) && parseInt(value, 10) > 0) {
    return parseInt(value, 10);
  }
  return fallback;
}

function percentile(sorted: number[], p: number): number {
  if (sorted.length === 0) return 0;
  const idx = Math.min(Math.floor((p * sorted.length) / 100), sorted.length - 1);
  return sorted[idx];
}

async function fetchOnce(url: string): Promise<{ elapsed: number; ok: boolean }> {
  const start = performance.now();
  try {
    const res = await fetch(url);
    await res.arrayBuffer();
    return { elapsed: performance.now() - start, ok: res.ok };
  } catch {
    return { elapsed: performance.now() - start, ok: false };
  }
}

async function runBatch(
  url: string,
  count: number
): Promise<{ elapsed: number; ok: boolean }[]> {
  return Promise.all(Array.from({ length: count }, () => fetchOnce(url)));
}

async function main(): Promise<void> {
  const url = process.env.BENCH_URL ?? DEFAULT_URL;
  const total = envInt("BENCH_REQUESTS", DEFAULT_REQUESTS);
  const concurrency = envInt("BENCH_CONCURRENCY", DEFAULT_CONCURRENCY);

  console.log("TypeScript benchmark client");
  console.log(`  Target:      ${url}`);
  console.log(`  Requests:    ${total}`);
  console.log(`  Concurrency: ${concurrency}\n`);

  try {
    const warmup = await fetch(url);
    if (!warmup.ok) throw new Error(`HTTP ${warmup.status}`);
  } catch (err) {
    console.error(`Server not reachable at ${url}:`, err);
    process.exit(1);
  }

  const latencies: number[] = [];
  let okCount = 0;
  let failed = 0;

  const start = performance.now();
  const batches = Math.ceil(total / concurrency);
  let sent = 0;

  for (let b = 0; b < batches; b++) {
    const batchSize = Math.min(concurrency, total - sent);
    const results = await runBatch(url, batchSize);
    for (const result of results) {
      latencies.push(result.elapsed);
      if (result.ok) okCount++;
      else failed++;
    }
    sent += batchSize;
  }
  const duration = (performance.now() - start) / 1000;

  latencies.sort((a, b) => a - b);
  const rps = total / duration;
  const avg = latencies.reduce((a, b) => a + b, 0) / latencies.length;

  console.log("Results");
  console.log("-------");
  console.log("Language:     TypeScript (Express client)");
  console.log(`URL:          ${url}`);
  console.log(`Duration:     ${duration.toFixed(3)}s`);
  console.log(`Requests/sec: ${rps.toFixed(2)}`);
  console.log(`Successful:   ${okCount}`);
  console.log(`Failed:       ${failed}`);
  console.log(`Avg latency:  ${avg.toFixed(2)}ms`);
  console.log(`p50 latency:  ${percentile(latencies, 50).toFixed(2)}ms`);
  console.log(`p95 latency:  ${percentile(latencies, 95).toFixed(2)}ms`);
  console.log(`p99 latency:  ${percentile(latencies, 99).toFixed(2)}ms`);
}

main();
